Bluetooth: Samples: Add hci_uart_async

alwa-nordic opened the pull request:
This sample is an alternative implementation of hci_uart. The new sample differs from the existing sample in that it uses the async UART API instead of the interrupt driven API.

Included in this commit is a new test for HCI UART flow control. It's enabled for hci_uart_async. The test can also excercise the existing hci_uart sample (with some minimal changes to allow compiling in the mock controller and test suite). The existing hci_uart sample currently fails this flow control test.

alwa-nordic committed:
bluetooth: samples: Add hci_uart_async

This sample is an alternative implementation of hci_uart. The new sample
differs from the existing sample in that it uses the async UART API
instead of the interrupt driven API.

Included in this commit is a new test for HCI UART flow control. It's
enabled for hci_uart_async. The test can excercise also the existing
hci_uart sample (with some minimal changes to allow compiling in the
mock controller and test suite). The existing hci_uart sample currently
fails the flow control test.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

alwa-nordic commented on samples/bluetooth/hci_uart_async/app.overlay:
```diff
@@ -0,0 +1,31 @@
+        };
+};
+
+/* Some boards are by default assinging the &uart0 to these other
```
```suggestion
/* Some boards are by default assigning the &uart0 to these other
```

jhedberg commented on tests/samples/bluetooth/hci_uart_async/CMakeLists.txt:
Why are you adding a completely new `tests/samples/` directory? We already have `samples/`and `tests/`, so adding a `samples` directory under `tests` seems quite confusing and unnecessary.

alwa-nordic replied:
I don't know where to put the test. My thinking was that since the DUT has path `samples/bluetooth/hci_uart_async`, the test should start with the same but with `tests/` prepended, ergo `tests/samples/bluetooth/hci_uart_async`. Possibly with sub-directories under that when we get more test.

I considered if the test should be co-located with the implementation. One could argue that the sample folder is like an independent application project folder, and it would make sense to have the tests there. But I was unsure if CI twister scans for tests outside `/tests`. And of course this would break with the pattern that all tests so far have been in `/tests`.

I considered putting the test under `/tests/bluetooth/samples/hci_uart_async`. But I found it confusing myself when vscode abbreviates my open folders to (the shortest unique suffixes) `bluetooth/hci_uart_async` (the sample) and `samples/hci_uart_async` (the test). If that was confusing to you, that's because it is.

RFC.

jori-nordic replied:
Could you elaborate why it's confusing @jhedberg ? 
It seems simple to me: `tests/samples` -> tests for the samples.

We could put it under `tests/bluetooth/samples/hci_uart_async` but that's more confusing imo because it flips the  path around compared to where the sample lives `samples/bluetooth/hci_uart_async`.

jhedberg replied:
> It seems simple to me: `tests/samples` -> tests for the samples.

What kind of thought process do you go through to arrive with the word "for" to associate the parent and child directory together? I don't see how that's the obvious meaning that we can assume everyone to immediately understand when they see it. When I look at the Zephyr top level directory, I see "tests" and I think "Ok, there are tests under this directory". When I see "samples" I think "Ok, there are samples under this directory". Now, when I'm in "tests" and I see "samples", I think "Hold on, I was told there are tests here, but now suddenly you're telling me there are samples!?" :) 

Either way, introducing a new top-level directory with a generic name ("samples") under `tests/` is not something we can decide on strictly within a Bluetooth scope.

> We could put it under `tests/bluetooth/samples/hci_uart_async` but that's more confusing imo because it flips the path around compared to where the sample lives `samples/bluetooth/hci_uart_async`.

What's wrong with using `tests/bluetooth/hci_uart_async` for the test?

alwa-nordic replied:
I'll move it to `tests/bluetooth/hci_uart_async`.

jori-nordic commented on samples/bluetooth/hci_uart_async/app.overlay:
```diff
@@ -0,0 +1,31 @@
+ * there is a board-specific overlay in `./boards`.
+ *
+ * Most boards define a convenient `&uart0`. It's used here to make the
+ * sample 'just work' automatically for those boards.
```
just works

alwa-nordic replied:
You mean like below? I think it sounds worse.

> It's used here to make the sample 'just works' automatically for those boards.

jori-nordic replied:
Huh, maybe the diff changed when I commented. Agree it makes more sense as-is.

jori-nordic commented on samples/bluetooth/hci_uart_async/prj.conf:
```diff
@@ -0,0 +1,25 @@
+CONFIG_BT_BUF_EVT_DISCARDABLE_SIZE=255
+
+# Send an initial HCI_Command_Complete event on boot without waiting for
+# HCI_Reset.
```
can you add: `Needs to be enabled in the host application too.`

alwa-nordic replied:
I'll add it. It needs to because it will log errors in the Host otherwise right?

There is wording in the spec that suggests that the Host must always accept these.

> The Controller can send an HCI_Command_Complete event with Command Opcode 0x0000 at any time [...]
>
> -- Core 5.4 E 7.7.14

jori-nordic replied:
yeah it will log a warning I think. What I wanna say is that this config has to be set on both MCUs for the feature to work properly (waiting for the NOP on bt_init).

jhedberg reviewed (changes requested):
Please split the non-Bluetooth changes (to `serial_test.c`, `serial_test.h`, and `vnd,serial.yaml`) into its own commit

alwa-nordic committed:
dts: bindings: Remove `reg` from `vnd,serial`

`vnd,serial` is a virtual device which does not have an address.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

alwa-nordic committed:
drivers: serial_test: Implement interrupt and async APIs

This allows testing code that uses the UART interrupt-driven API and
UART asynchronous API.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

dcpleung reviewed (approved):
LGTM on `serial_test.c`.

dcpleung reviewed (changes requested):
Almost forgot... could you add to `serial_test.c` the same as https://github.com/zephyrproject-rtos/zephyr/pull/60396?

alwa-nordic commented:
> Almost forgot... could you add to `serial_test.c` the same as #60396?

Done. But wouldn't it be better to have that enforced centrally in `uart_irq_callback_user_data_set` and `uart_callback_set`? I.e. Both of those functions would call the other with NULL (to clear the other callback) if it itself did not get NULL (to clear the callback).

dcpleung commented:
> > Almost forgot... could you add to `serial_test.c` the same as #60396?
> 
> Done. But wouldn't it be better to have that enforced centrally in `uart_irq_callback_user_data_set` and `uart_callback_set`? I.e. Both of those functions would call the other with NULL (to clear the other callback) if it itself did not get NULL (to clear the callback).

That is done to allow more flexibility on the driver side. Some drivers may decide to disable certain interrupts when there is no need to do callback, which might interfere with the other API. The driver would know what needs to be done correctly, and that the current implementation let the driver decide.

dcpleung commented:
Oh... you add the callback changes to the bluetooth sample commit. Could you move those to the commit which changes serial_test.c?

alwa-nordic commented:
> Oh... you add the callback changes to the bluetooth sample commit. Could you move those to the commit which changes serial_test.c?

Done. Please re-review :)

dcpleung reviewed (dismissed):
LGTM to me on `serial_test.c`.

jori-nordic reviewed (approved):
Could you test it one last time before merging?

carlescufi commented:
DNM until @nordic-krch has had a chance to review this please

nordic-krch commented on samples/bluetooth/hci_uart_async/src/hci_uart_async.c:
```diff
@@ -0,0 +1,371 @@
+		K_POLL_EVENT_INITIALIZER(K_POLL_TYPE_SIGNAL, K_POLL_MODE_NOTIFY_ONLY, sig),
+	};
+
+	/* Maybe take a mutex? */
```
would be good to answer that question.

alwa-nordic replied:
I marked this function not-thread-safe. It's only invoked from one thread, so there is no need for a mutex.

nordic-krch commented on samples/bluetooth/hci_uart_async/src/hci_uart_async.c:
```diff
@@ -0,0 +1,371 @@
+	if (evt->type == UART_RX_DISABLED) {
+		k_poll_signal_raise(&uart_h2c_rx_sig, 0);
+	}
+	if (evt->type == UART_RX_STOPPED) {
```
you can use `else` since only one event arrives at time.

Minor optimization: what about setting signal variable and variable in the `if` and then calling `k_poll_signal_raise` only once (should be less code).

nordic-krch commented on samples/bluetooth/hci_uart_async/src/hci_uart_async.c:
```diff
@@ -0,0 +1,371 @@
+} __packed cc_evt = {
+	.h4 = H4_EVT,
+	.hdr = {.evt = BT_HCI_EVT_CMD_COMPLETE, .len = sizeof(struct bt_hci_evt_cmd_complete)},
+	.cc = {.ncmd = 1, /* TODO: Should this value come from a config? */
```
would be good to get rid of `TODO`

alwa-nordic replied:
I removed the TODO. It was like this in the original `hci_uart` sample. I assume `CONFIG_BT_WAIT_NOP` wants it like this.

nordic-krch commented on samples/bluetooth/hci_uart_async/src/hci_uart_async.c:
```diff
@@ -0,0 +1,371 @@
+			}
+		}
+
+		k_yield();
```
is this needed? Wouldn't you just block this thread on waiting for receiving next packet?

alwa-nordic replied:
It's there in the original `hci_uart` sample.

https://github.com/zephyrproject-rtos/zephyr/blob/168de28b40a4ed7bcd4da8c59d74e2341904fff0/samples/bluetooth/hci_uart/src/main.c#L264-L267

Presumably it gives fairer scheduling. It may be that there is no [reschedule point] in e.g. `k_poll` if it returns without waiting.

[reschedule point]: https://docs.zephyrproject.org/latest/kernel/services/scheduling/index.html#concepts

nordic-krch reviewed (commented):
Looks ok. Few minor comments.
Have you consider making a common h4 layer for host and controller? I've done something like that in the past but it didn't get merged: https://github.com/zephyrproject-rtos/zephyr/pull/26692

nordic-krch commented on samples/bluetooth/hci_uart_async/src/hci_uart_async.c:
```diff
@@ -0,0 +1,405 @@
+				return;
+			}
+			LOG_HEXDUMP_DBG(payload_dst, payload_size, "h2c: hci payload");
+		} else {
+			/* Discard oversize packet. */
+			uint8_t *discard_dst;
+			uint16_t discard_size;
+
+			LOG_WRN("h2c: Discarding oversize h4_type %d payload_size %d.", h4_type,
+				payload_size);
+
+			/* Reset `buf` so all of it is available. */
+			net_buf_reset(buf);
+			discard_dst = net_buf_tail(buf);
+			discard_size = net_buf_max_len(buf);
+
+			while (payload_size) {
+				uint16_t read_size = MIN(payload_size, discard_size);
+
+				err = uart_h2c_rx(discard_dst, read_size);
+				if (err) {
+					net_buf_unref(buf);
+					return;
+				}
+
+				payload_size -= read_size;
+			}
+
+			net_buf_unref(buf);
+			buf = NULL;
+		}
```
I'm not the author and also don't know that.

cvinayak reviewed (dismissed):
Add sample.yaml file with supported platforms
